flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskLocalStateStoreImpl.java

```diff
@@ -141,6 +158,29 @@ public void storeLocalState(
                     subtaskIndex);
         }
 
+        for (Map.Entry<OperatorID, OperatorSubtaskState> subtaskStateEntry :
+                localState.getSubtaskStateMappings()) {
+            for (KeyedStateHandle keyedStateHandle :
+                    subtaskStateEntry.getValue().getManagedKeyedState()) {
+                if (keyedStateHandle instanceof ChangelogStateBackendHandle) {
+                    ChangelogStateBackendHandle changelogStateBackendHandle =
+                            (ChangelogStateBackendHandle) keyedStateHandle;
+                    long materializationID = changelogStateBackendHandle.getMaterializationID();
+                    if (currentMaterializationID == null
+                            || materializationID != currentMaterializationID.f0) {
+                        currentMaterializationID = Tuple2.of(materializationID, checkpointId);
+                        referredByCheckpointID.clear();
```

Comment:
Could you clarify what happens if two snapshots with different materialization IDs complete out of order?
Isn't it possible that an older checkpoint (say, CP1 with mID=1) will "override" the newer one (say, CP2 with mID=2)?

Reply:
I add `lastCheckpointId` to check whether checkpoint is out of order.
- if yes, don't write to local store, return directly. According to the materialization of the out-of-order checkpoint, there are two cases:
     - if cp with a newer materialization, say, (1)cp1 with m1, (2) cp3 with m2, (3) cp2 with m1.  when confirming cp3, cp1 and m1 would be discarded, and cp2 would not write to local store. therefore, no additional checkpoints are kept apart from cp3 and m2.
     -  if cp without a newer materialization, say, (1)cp1 with m1, (2) cp3 with m1, (3) cp2 with m1. when confirming cp3, cp1 would retain, and cp2 would not write to local store. finally, cp1,m1,cp3 are kept.